pkg/reconciler/extension/extension.go

```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package extension
+
+import (
+	"context"
+
+	autoscalingv1alpha1 "knative.dev/serving/pkg/apis/autoscaling/v1alpha1"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+// ServingExtension enables pluggable extensive features added into the controller
+type ServingExtension interface {
+	// TransformRevision transforms the existing revision.
+	TransformRevision(*v1.Revision) *v1.Revision
+	// PostRevisionReconcile implements a custom reconcile loop the after the revision is created.
+	PostRevisionReconcile(context.Context, *autoscalingv1alpha1.PodAutoscaler, *autoscalingv1alpha1.PodAutoscaler) error
+	// PostConfigurationReconcile implements a custom reconcile loop after the configuration is created.
+	PostConfigurationReconcile(context.Context, *v1.Service, *v1.Configuration) error
+	// TransformService transforms the existing service.
+	TransformService(*v1.Service) *v1.Service
+	// UpdateExtensionStatus updates the status of the custom resources in the extension.
+	UpdateExtensionStatus(context.Context, *v1.Service) (bool, error)
```

Comment:
Serving controllers shouldn't be responsible for updating the status of a different resource that it doesn't own. I'd imagine you would watch the resources you care about in your own controller and update the status there.

Reply:
Perhaps I did not name this function well. 
This is what I am planning to do: in the end, the controller gets the information about how ready the extension's CR. If the extension's CR is ready as true, the controller here knows everything is ready; if not, the controller here will kick-off the reconcile again after a certain time(that's the reason I run a re-enque), and check the status again, until the extension's CR is ready.

Reply:
I thought we discussed in the Serving WG that you would create a sibling resource to Service that would manage a route and configuration. So I would imagine you don't need to incorporate this extension into the serving repo.